### PR TITLE
Automatically use torch.jit.script for all operations

### PR DESCRIPTION
**Pull request by Luthaf**
This should help with #482, and torch script compatibility was already a requirement anyway.

The diff is a bit larger than I'd like because when using 

```
@torch.jit.script
def foo():
    ...
```

all functions called inside `foo` must already be defined.

As a small side change, I renamed `_classes` to `_backend` which feels more descriptive.

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [ ] ~Documentation updated (for new features)?~
 - [x] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://metatensor--504.org.readthedocs.build/en/504/

### CODE REVIEW

**Review by frostedoyster — COMMENTED** [review]
Is there a way to use the non-scripted classes? (E.g. for better error messages?)
EDIT: E.g. "re-run with METATENSOR_OPERATIONS_TORCHSCRIPT=0 for a better error message"

**Comment by agoscinski** [comment]
```
Is there a way to use the non-scripted classes? (E.g. for better error messages?)
EDIT: E.g. "re-run with METATENSOR_OPERATIONS_TORCHSCRIPT=0 for a better error message
```
Maybe this can be put into CONTRIBUTING.rst into the `Running tests`? Otherwise looks good. PR seems very simple, just creates a lot of line noise because of change of order of the functions which is needed for torch script

File: python/metatensor-operations/metatensor/operations/_backend.py [hunk 1]
```diff
@@ -16,12 +16,17 @@ def torch_jit_is_scripting():
     return False
 
 
+def torch_jit_script(function):
+    return function
```
Comment (PicoCentauri) [inline python/metatensor-operations/metatensor/operations/_backend.py:20]: Maybe we should a small comment why we need this function. On a first glance this just looks like a useless unity operation.

File: python/metatensor-operations/metatensor/operations/reduce_over_samples.py [hunk 2]
```diff
@@ -437,12 +428,15 @@ def sum_over_samples_block(
     [[ 4  7 10]
      [17 19 21]]
     """
+    if isinstance(sample_names, str):
+        sample_names = [sample_names]
```
Comment (PicoCentauri) [inline python/metatensor-operations/metatensor/operations/reduce_over_samples.py:432]: This conversion done in the private functions? I think we can skip this here or do we need this due to torchscript?

**Review by PicoCentauri — COMMENTED** [review]
Two minor questions. Very good test cleanup as well.

**Comment by Luthaf** [comment]
> Is there a way to use the non-scripted classes? (E.g. for better error messages?)

I think we can use the normal PyTorch mechanism for this: https://pytorch.org/docs/stable/jit.html#disable-jit-for-debugging

I'll add a line to Contributing.rst

File: python/metatensor-operations/metatensor/operations/_backend.py (hunk 1 above)
Comment (Luthaf) [reply python/metatensor-operations/metatensor/operations/_backend.py:20]: Good point

File: python/metatensor-operations/metatensor/operations/reduce_over_samples.py (hunk 2 above)
Comment (Luthaf) [reply python/metatensor-operations/metatensor/operations/reduce_over_samples.py:432]: It is not: the private function want `List[str]`, not `Union[List[str], str]`. I can change them to do this though!

Comment (PicoCentauri) [reply python/metatensor-operations/metatensor/operations/reduce_over_samples.py:432]: Adding this to the private functions looks for consistent to me. The public functions are just as wrappers for the private ones setting the parameters to the private ones. Sorry for the extra work.

**Review by frostedoyster — APPROVED** [review]

**Pull request merged**